library(Battenberg)
library(doParallel)

###############################################################################
# 2015-04-12
# A single core Battenberg v2.0 SNP6 pipeline implementation.
# sd11@sanger.ac.uk
###############################################################################

# Sample specific
TUMOURNAME = "A46B_b"
IS.MALE = F
TUMOURCEL = "/lustre/scratch112/sanger/cancer_external/DBGap/TCGA_phs000178.v8.p7/CELfiles/PRAD/RIELS_p_TCGA_268_269_N_GenomeWideSNP_6_G07_1306434.CEL"
NORMALCEL = "/lustre/scratch112/sanger/cancer_external/DBGap/TCGA_phs000178.v8.p7/CELfiles/PRAD/RIELS_p_TCGA_268_269_N_GenomeWideSNP_6_H05_1306362.CEL"
RUN_DIR = "/lustre/scratch110/sanger/sd11/battenberg_package_test/NASCR-0016_bb_v2.0_singlecore"

# Parallelism parameters
NTHREADS = 6

# General static
IMPUTEINFOFILE = "/nfs/users/nfs_s/sd11/repo/battenberg/impute_info.txt"
G1000PREFIX = "/lustre/scratch110/sanger/sd11/Documents/GenomeFiles/battenberg_1000genomesloci2012/1000genomesAlleles2012_chr"

# General SNP6 specific
PROBLEMLOCI = NA
SNP6_REF_INFO_FILE = "/lustre/scratch110/sanger/sd11/Documents/GenomeFiles/battenberg_snp6/snp6_ref_info_file.txt"
APT_PROBESET_GENOTYPE_EXE = "apt-probeset-genotype"
APT_PROBESET_SUMMARIZE_EXE = "apt-probeset-summarize"
NORM_GENO_CLUST_EXE = "normalize_affy_geno_cluster.pl"
BIRDSEED_REPORT_FILE = "birdseed.report.txt" # No control over the name of this file, as it is automatically generated by APT within cel2baf.logr

# Parameters
PLATFORM_GAMMA = 0.55
PHASING_GAMMA = 1
SEGMENTATION_GAMMA = 10
CLONALITY_DIST_METRIC = 0
ASCAT_DIST_METRIC = 1
MIN_PLOIDY = 1.6
MAX_PLOIDY = 4.8
MIN_RHO = 0.1
MIN_GOODNESS_OF_FIT = 0.63
BALANCED_THRESHOLD = 0.51
MIN_NORMAL_DEPTH = 10

# Change to work directory and load the chromosome information
setwd(RUN_DIR)
chrom_names = get.chrom.names(IMPUTEINFOFILE, IS.MALE)

# Setup for parallel computing
clp = makeCluster(NTHREADS)
registerDoParallel(clp)

# Extract the LogR and BAF from both tumour and normal cel files.
cel2baf.logr(normal_cel_file=NORMALCEL, 
             tumour_cel_file=TUMOURCEL, 
             output_file=paste(TUMOURNAME, "_lrr_baf.txt", sep=""), 
             snp6_reference_info_file=SNP6_REF_INFO_FILE, 
             apt.probeset.genotype.exe=APT_PROBESET_GENOTYPE_EXE, 
             apt.probeset.summarize.exe=APT_PROBESET_SUMMARIZE_EXE, 
             norm.geno.clust.exe=NORM_GENO_CLUST_EXE)

gc.correct(samplename=TUMOURNAME, 
           infile.logr.baf=paste(TUMOURNAME, "_lrr_baf.txt", sep=""), 
           outfile.tumor.LogR=paste(TUMOURNAME, ".tumour.LogR.txt", sep=""), 
           outfile.tumor.BAF=paste(TUMOURNAME, ".tumour.BAF.txt", sep=""), 
           outfile.normal.LogR=paste(TUMOURNAME, ".normal.LogR.txt", sep=""), 
           outfile.normal.BAF=paste(TUMOURNAME, ".normal.BAF.txt", sep=""), 
           outfile.probeBAF=paste(TUMOURNAME, "_probeBAF.txt", sep=""),
           snp6_reference_info_file=SNP6_REF_INFO_FILE,
           birdseed_report_file=BIRDSEED_REPORT_FILE)

# for (chrom in 1:length(chrom_names)) {
#, .export=c("generate.impute.input.snp6","run.impute","combine.impute.output","GetChromosomeBAFs_SNP6","plot.haplotype.data")
foreach(i=1:length(chrom_names)) %dopar% {
  # Transform input into a format that Impute2 takes
  generate.impute.input.snp6(infile.probeBAF=paste(TUMOURNAME, "_probeBAF.txt", sep=""), 
     outFileStart=paste(TUMOURNAME, "_impute_input_chr", sep=""), 
     chrom=chrom, 
     chr_names=chrom_names,
     problemLociFile=PROBLEMLOCI, 
     snp6_reference_info_file=SNP6_REF_INFO_FILE, 
     imputeinfofile=IMPUTEINFOFILE,
     is.male=IS.MALE,
     heterozygousFilter="none")

  # Run impute on the files
  run.impute(inputfile=paste(TUMOURNAME, "_impute_input_chr", chrom, ".txt", sep=""),
    outputfile.prefix=paste(TUMOURNAME, "_impute_output_chr", chrom, ".txt", sep=""),
    is.male=IS.MALE,
    imputeinfofile=IMPUTEINFOFILE,
    chrom=chrom)
                 
  # As impute runs in windows across a chromosome we need to assemble the output
  combine.impute.output(inputfile.prefix=paste(TUMOURNAME, "_impute_output_chr", chrom, ".txt", sep=""),
    outputfile=paste(TUMOURNAME, "_impute_output_chr", chrom, "_allHaplotypeInfo.txt", sep=""),
    is.male=IS.MALE,
    imputeinfofile=IMPUTEINFOFILE,
    region.size=5000000,
    chrom=chrom)
  
  # Transform the impute output into haplotyped BAFs
  GetChromosomeBAFs_SNP6(chrom=chrom,
    alleleFreqFile=paste(TUMOURNAME, "_impute_input_chr", chrom, "_withAlleleFreq.csv", sep=""),
    haplotypeFile=paste(TUMOURNAME, "_impute_output_chr", chrom, "_allHaplotypeInfo.txt", sep=""),
    samplename=TUMOURNAME,
    outputfile=paste(TUMOURNAME, "_chr", chrom, "_heterozygousMutBAFs_haplotyped.txt", sep=""),
    chr_names=chrom_names)
                                                                                           
  # Plot what we have until this point
  plot.haplotype.data(haplotyped.baf.file=paste(TUMOURNAME, "_chr", chrom, "_heterozygousMutBAFs_haplotyped.txt", sep=""),
     imageFileName=paste(TUMOURNAME,"_chr",chr_name,"_heterozygousData.png",sep=""), 
     samplename=TUMOURNAME, 
     chrom=chrom, 
     chr_names=chrom_names)
  
  # Cleanup temp Impute output
  cmd = paste("rm ", paste(TUMOURNAME, "_impute_input_chr", sep=""), chrom, "_*_*.*", sep="")
  system(cmd)
}

# Kill the threads as from here its all single core
stopCluster(clp)
                                                                                         
# Combine all the BAF output into a single file
combine.baf.files(inputfile.prefix=paste(TUMOURNAME, "_chr", sep=""), 
       inputfile.postfix="_heterozygousMutBAFs_haplotyped.txt", 
       outputfile=paste(TUMOURNAME, "_heterozygousMutBAFs_haplotyped.txt", sep=""),
       no.chrs=length(chrom_names),
       gamma=10,
       phasegamma=3,
       kmin=3,
       phasekmin=3)

# Segment the phased and haplotyped BAF data
segment.baf.phased(inputfile=paste(TUMOURNAME, "_heterozygousMutBAFs_haplotyped.txt", sep=""), 
        outputfile=paste(TUMOURNAME, ".BAFsegmented.txt", sep=""))

# Fit a clonal copy number profile
fit.copy.number(samplename=TUMOURNAME,
     outputfile.prefix=paste(TUMOURNAME, "_", sep=""),
     inputfile.baf.segmented=paste(TUMOURNAME, ".BAFsegmented.txt", sep=""), 
     inputfile.baf=paste(TUMOURNAME,"_mutantBAF.tab", sep=""), 
     inputfile.logr=paste(TUMOURNAME,"_mutantLogR.tab", sep=""), 
     dist_choice=CLONALITY_DIST_METRIC, 
     ascat_dist_choice=ASCAT_DIST_METRIC, 
     min.ploidy=MIN_PLOIDY, 
     max.ploidy=MAX_PLOIDY, 
     min.rho=MIN_RHO, 
     min.goodness=MIN_GOODNESS_OF_FIT, 
     uninformative_BAF_threshold=BALANCED_THRESHOLD, 
     gamma_param=PLATFORM_GAMMA, 
     use_preset_rho_psi=F, 
     preset_rho=NA, 
     preset_psi=NA, 
     read_depth=30)

# Go over all segments, determine which segements are a mixture of two states and fit a second CN state
callSubclones(sample.name=TUMOURNAME, 
   baf.segmented.file=paste(TUMOURNAME, ".BAFsegmented.txt", sep=""), 
   logr.file=paste(TUMOURNAME,"_mutantLogR.tab", sep=""), 
   rho.psi.file=paste(TUMOURNAME, "_rho_and_psi.txt",sep=""), 
   output.file=paste(TUMOURNAME,"_subclones.txt", sep=""), 
   output.figures.prefix=paste(TUMOURNAME,"_subclones_chr", sep=""), 
   chr_names=chrom_names, 
   gamma=PLATFORM_GAMMA, 
   segmentation.gamma=NA, 
   siglevel=0.05, 
   maxdist=0.01, 
   noperms=1000)